python/ray/data/_internal/execution/interfaces.py

```diff
@@ -0,0 +1,267 @@
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Iterator, Tuple
+
+import ray
+from ray.data._internal.stats import DatasetStats, StatsDict
+from ray.data.block import Block, BlockMetadata
+from ray.types import ObjectRef
+
+
+@dataclass
+class RefBundle:
+    """A group of data block references and their metadata.
+
+    Operators take in and produce streams of RefBundles.
+
+    Most commonly an RefBundle consists of a single block object reference.
+    In some cases, e.g., due to block splitting, or for a SortReduce task, there may
+    be more than one block.
+
+    Block bundles have ownership semantics, i.e., shared_ptr vs unique_ptr. This
+    allows operators to know whether they can destroy blocks when they don't need
+    them. Destroying blocks eagerly is more efficient than waiting for Python GC /
+    Ray reference counting to kick in.
+    """
+
+    # The size_bytes must be known in the metadata, num_rows is optional.
+    blocks: List[Tuple[ObjectRef[Block], BlockMetadata]]
+
+    # Whether we own the blocks (can safely destroy them).
+    owns_blocks: bool
+
+    def __post_init__(self):
+        for b in self.blocks:
+            assert isinstance(b, tuple), b
+            assert len(b) == 2, b
+            assert isinstance(b[0], ray.ObjectRef), b
+            assert isinstance(b[1], BlockMetadata), b
+            if b[1].size_bytes is None:
+                raise ValueError(
+                    "The size in bytes of the block must be known: {}".format(b)
+                )
+
+    def num_rows(self) -> Optional[int]:
+        """Number of rows present in this bundle, if known."""
+        total = 0
+        for b in self.blocks:
+            if b[1].num_rows is None:
+                return None
+            else:
+                total += b[1].num_rows
+        return total
+
+    def size_bytes(self) -> int:
+        """Size of the blocks of this bundle in bytes."""
+        return sum(b[1].size_bytes for b in self.blocks)
+
+    def destroy_if_owned(self) -> int:
+        """Clears the object store memory for these blocks if owned.
+
+        Returns:
+            The number of bytes freed.
+        """
+        raise NotImplementedError
+
+
+@dataclass
+class ExecutionOptions:
+    """Common options that should be supported by all Executor implementations."""
+
+    # Max number of in flight tasks. This is a soft limit.
+    parallelism_limit: Optional[int] = None
+
+    # Example: set to 1GB and executor will try to limit object store
+    # memory usage to 1GB. This is a soft limit.
+    memory_limit_bytes: Optional[int] = None
```

Comment:
wondering how could bulk executor support this option? 

I know we declare it as a soft limit. But isn't it that the current bulk execution does not respect memory limit at all?

Should this be an option only for pipelined executor?

Reply:
Yeah, added a note these options are only supported for streaming mode.